X11R6/contrib/programs/xcolorize/xcolorize.man
.\" $Id: xcolorize.man,v 6.1 94/05/25 13:53:50 dd Exp $
.\" Copyright (c) 1994 David DiGiacomo
.\"
.\" The X Consortium, and any party obtaining a copy of these files from
.\" the X Consortium, directly or indirectly, is granted, free of charge, a
.\" full and unrestricted irrevocable, world-wide, paid up, royalty-free,
.\" nonexclusive right and license to deal in this software and
.\" documentation files (the "Software"), including without limitation the
.\" rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons who receive
.\" copies from any such party to do so.  This license includes without
.\" limitation a license to do the foregoing actions under any patents of
.\" the party supplying this software to the X Consortium.
.TH XCOLORIZE 1 "Release 6" "X Version 11"
.SH NAME
xcolorize \- start X client with pleasant background color
.SH SYNOPSIS
.B xcolorize
[\fB\-help\fP]
[\fB\-dark\fP]
[\fB\-only\fP]
[\fB\-rv\fP]
[\fB\-old\fP|\fB\-oldsh\fP|\fB\-sun\fP]
[\fIcommand\fP [\fIargs ...\fP]]
.SH DESCRIPTION
The
.I xcolorize
program is used to give an X client program 
a pleasant randomly generated background color.
It inserts the appropriate
.B \-fg
and
.B \-bg
options in the client command line.
.SH OPTIONS
.TP 8
.B \-help
Prints a help message.
.TP 8
.B \-dark
Generates a dark color background and white foreground instead of the
default pastel color background and black foreground.
.TP 8
.B \-only
Generates only a random color specification instead of complete
foreground and background color options.
.TP 8
.B \-rv
Reverses the generated foreground and background colors.
.TP 8
.B \-old
Generates the old #RRGGBB syntax instead of the newer rgb:RR/GG/BB syntax.
Useful for pre-X11R5 clients which don't understand the new syntax.
.TP 8
.B \-oldsh
Same as \fB\-old\fP, but prefixes the \fB#\fP character with a backslash
to prevent it from being treated as a shell comment character.  This is
sometimes useful when the X client being colorized is implemented as a
shell script.
.TP 8
.B \-sun
Generates \fIXView\fP/\fISunView\fP decimal RGB syntax.  Useful when
colorizing \fISunView\fP and some \fIXView\fP clients which don't
understand X color syntax.
.TP 8
\fIcommand\fP [\fIargs ...\fP]
The generated color options are inserted
between the command and the first argument supplied (if any).  If no
command is given, the generated color options are written to stdout.
.SH EXAMPLE
Below is an example
.I .xinitrc
file which uses
.I xcolorize
to make the screen more attractive:
.sp
.in 1.0i
.nf
xrdb \-load $HOME/.Xresources
xcolorize -dark xterm \-C \-title Console \-geometry 80x24+0+0 &
xsetroot \-solid `xcolorize -dark -only`
xcolorize xclock \-geometry 50x50\-0+0 \-bw 0 &
xcolorize -dark xload \-geometry 50x50\-50+0 \-bw 0 &
xcolorize xterm \-geometry 80x24+0\-0 &
twm
.fi
.in -1.0i
.sp
.SH "SEE ALSO"
.IR X (1)
.SH AUTHOR
David DiGiacomo, david@slack.com
.SH BUGS
The random color generation algorithm is not very good.
.sp 1
There should be a way to generate both foreground and background colors
(non-clashing).
.sp 1
If you colorize too many applications you'll fill up your colormap.


X11R6/contrib/programs/xcolorize/xcolorize.sh
#! /bin/sh
# $Id: xcolorize.sh,v 6.1 94/05/25 13:53:51 dd Exp $

# Copyright (c) 1994 David DiGiacomo david@slack.com
#
# The X Consortium, and any party obtaining a copy of these files from
# the X Consortium, directly or indirectly, is granted, free of charge, a
# full and unrestricted irrevocable, world-wide, paid up, royalty-free,
# nonexclusive right and license to deal in this software and
# documentation files (the "Software"), including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons who receive
# copies from any such party to do so.  This license includes without
# limitation a license to do the foregoing actions under any patents of
# the party supplying this software to the X Consortium.

# xcolorize:
#	start X client program with pleasant background color

cfmt=xrgb	# default to rgb:rr/gg/bb format
dark=0		# default to light (pastel) colors
rv=0		# default to non-reverse video
coloronly=0	# default to printing window fg/bg arguments

while [ $# -gt 0 ]
do
	case "$1" in
	-rgb)		cfmt=xrgb shift ;;
	-on*|-c*)	coloronly=1 shift ;;
	-o*sh|-sh)	cfmt=xoldsh shift ;;
	-o*|-\#*)	cfmt=xold shift ;;
	-s*|-xv*|-dec*)	cfmt=sun shift ;;
	-d*)		dark=1 shift ;;
	-p*)		dark=0 shift ;;
	-r*)		rv=1 shift ;;
	-h*)		echo 'usage:'
			echo "	`basename $0` [options] [command args...]"
			echo ''
			echo 'options:'
			echo ''
			echo '	-dark	choose a dark color'
			echo '	-pastel	choose a light color [default]'
			echo '	-only	generate color only (no -fg/-bg)'
			echo '	-rv	reverse foreground and background'
			echo ''
			echo '	-old	use #RRGGBB color format'
			echo '	-oldsh	use \#RRGGBB color format'
			echo '	-rgb	use rgb:RR/GG/BB color format [default]'
			echo '	-sun	use XView/SunView color format'
			echo ''
			echo '	-help	print this message'
			echo ''
			echo '	default command is echo '
			exit 0
			;;
	--)		shift; break;;
	-*)		echo 'usage:' 1>&2
			echo "`basename $0`" '[-help]' \
				'[-dark] [-only] [-rv]' \
				'[-old|-oldsh|-sun]' \
				'[command args...]' 1>&2
			exit 1 ;;
	*)		break ;;
	esac
done

# HSV to RGB conversion parameters
j=100	# max sv value
k=255	# max rgb value

fg= bg=
case $coloronly in
0)
	case $cfmt.$dark.$rv in
	sun.0.0) fg=-Wg bg=-Wb ;;
	sun.1.0) fg='-Wg -Wf 255 255 255' bg=-Wb ;;
	sun.0.1) fg='-Wg -Wb' bg='-Wf 255 255 255' ;;
	sun.1.1) fg='-Wg -Wb' bg='-Wf 0 0 0' ;;
	x*.0.0) fg='-fg black' bg='-bg' ;;
	x*.1.0) fg='-fg white' bg='-bg' ;;
	x*.0.1) fg='-bg black' bg='-fg' ;;
	x*.1.1) fg='-bg white' bg='-fg' ;;
	esac
	;;
esac

case $cfmt in
xrgb)	fmt="$fg $bg rgb:%02x/%02x/%02x" ;;
xold)	fmt="$fg $bg #%02x%02x%02x" ;;
xoldsh)	fmt="$fg $bg \#%02x%02x%02x" ;;
sun)	fmt="$fg $bg %d %d %d" ;;
esac

# Recipe:
# 1. mash together date and PID
# 2. stir bits until lumpy (a real random number generator would be better)
# 3. extract "random" hue, saturation and value
# 4. convert to RGB
# 5. format as needed
#
# In theory the colors could be generated directly in TekHVC space,
# but I wasn't able to get satisfactory results that way.

color=`date '+%y %j %H %M %S' 2> /dev/null |
awk "
{
	pid = $$
	j = $j
	k = $k
	dark = $dark
	rv = $rv
"'
	time = (($1 * 365) + $2) * 86400
	time %= 65536
	time += (($3 * 60) + $4 * 60) + $5

	r = time * 49.07692307 + pid * 29.47619075
	r %= 262144

	h = r % 360
	r /= 359.3
	if (dark) {
		s = r % 11 + 50
		r /= 13
		v = r % 11 + 50
	}
	else {
		s = r % 9 + 12
		v = 100
	}
	s = s * k / j
	if (s < 0) s = 0
	if (s > k) s = k
	v = v * k / j
	if (v < 0) v = 0
	if (v > k) v = k

	if (s == 0) {
		h = 0;
		r = g = b = int(v);
	}
	else {
		i = int(h / 60);
		f = h % 60;
		p = int(v * (k - s) / k);
		q = int(v * (k - s * f / 60) / k);
		t = int(v * (k - s * (60 - f) / 60) / k);

		if (i == 0) {
			r = v; g = t; b = p;
		}
		else if (i == 1) {
			r = q; g = v; b = p;
		}
		else if (i == 2) {
			r = p; g = v; b = t;
		}
		else if (i == 3) {
			r = p; g = q; b = v;
		}
		else if (i == 4) {
			r = t; g = p; b = v;
		}
		else if (i == 5) {
			r = v; g = p; b = q;
		}
	}
	printf "'"$fmt"'\n",r,g,b
	exit
}
'`

case $# in
0)
	exec echo $color
	;;
*)
	prog="$1" shift
	exec "$prog" $color ${1+"$@"}
	;;
esac
